cpu/sam0_common/periph/usbdev.c

```diff
@@ -0,0 +1,739 @@
+{
+    DEBUG("Disabling IN irq\n");
+    ep_reg->EPINTENCLR.reg = USB_DEVICE_EPINTENCLR_TRCPT1 |
+                             USB_DEVICE_EPINTENCLR_STALL1 |
+                             USB_DEVICE_EPINTENCLR_TRFAIL1;
+}
+
+static void _bank_set_address(usbdev_ep_t *ep)
+{
+    UsbDeviceDescBank *bank = _bank_from_ep(ep);
+
+    bank->ADDR.reg = (uint32_t)ep->buf;
+}
+
+static int _bank_set_size(usbdev_ep_t *ep)
+{
+    UsbDeviceDescBank *bank = _bank_from_ep(ep);
+    unsigned val = 0x00;
+
+    switch (ep->len) {
+        case 8:
+            val = 0x0;
+            break;
+        case 16:
+            val = 0x1;
+            break;
+        case 32:
+            val = 0x2;
+            break;
+        case 64:
+            val = 0x3;
+            break;
+        case 128:
+            val = 0x4;
+            break;
+        case 256:
+            val = 0x5;
+            break;
+        case 512:
+            val = 0x6;
+            break;
+        case 1023:
+            val = 0x7;
+            break;
+        default:
+            return -1;
+    }
+    bank->PCKSIZE.bit.SIZE = val;
+    return val;
+}
+
+/**
+ * Checks if one of the endpoints interrupts flags for the OUT direction are
+ * set
+ */
+static bool _ep_out_flags_set(UsbDeviceEndpoint *ep_reg)
+{
+    return ep_reg->EPINTFLAG.reg  &
+           ep_reg->EPINTENSET.reg &
+           (USB_DEVICE_EPINTENSET_TRFAIL0 |
+            USB_DEVICE_EPINTENSET_TRCPT0 |
+            USB_DEVICE_EPINTENSET_RXSTP |
+            USB_DEVICE_EPINTENSET_STALL0);
+}
+
+/**
+ * Checks if one of the endpoints interrupts flags for the IN direction are
+ * set
+ */
+static bool _ep_in_flags_set(UsbDeviceEndpoint *ep_reg)
+{
+    return ep_reg->EPINTFLAG.reg &
+           ep_reg->EPINTENSET.reg &
+           (USB_DEVICE_EPINTENSET_TRFAIL1 |
+            USB_DEVICE_EPINTENSET_TRCPT1 |
+            USB_DEVICE_EPINTENSET_STALL1);
+}
+
+static void _set_address(sam0_common_usb_t *dev, uint8_t addr)
+{
+    dev->config->device->DADD.bit.DADD = addr;
+    /* Only enable the address if it is nonzero */
+    dev->config->device->DADD.bit.ADDEN = addr ? 1 : 0;
+}
+
+static bool _syncbusy_enable(sam0_common_usb_t *dev)
+{
+    return dev->config->device->SYNCBUSY.bit.ENABLE;
+}
+
+static bool _syncbusy_swrst(sam0_common_usb_t *dev)
+{
+    return dev->config->device->SYNCBUSY.bit.SWRST;
+}
+
+static void _gpio_connect_cb(void *arg)
+{
+    sam0_common_usb_t *usbdev = (sam0_common_usb_t*)arg;
+    usbdev->connect_change = true;
+    usbdev->usbdev.cb(&usbdev->usbdev, USBDEV_EVENT_ESR);
+}
+
+static inline void _poweron(void)
+{
+#if defined(CPU_FAM_SAMD21)
+    PM->AHBMASK.reg |= PM_AHBMASK_USB;
+    PM->APBBMASK.reg |= PM_APBBMASK_USB;
+    GCLK->CLKCTRL.reg = (uint32_t)(GCLK_CLKCTRL_CLKEN |
+                                   GCLK_CLKCTRL_GEN_GCLK0 |
+                                   (GCLK_CLKCTRL_ID(USB_GCLK_ID)));
+#elif defined(CPU_FAM_SAML21)
+    MCLK->AHBMASK.reg |= MCLK_AHBMASK_USB;
```

Comment:
You missed `MCLK->APBBMASK` for SAML21 family

Reply:
Thanks!